design_documents/cairo_deployment_scripts.md

```diff
@@ -0,0 +1,312 @@
+  - cons:
+    - boilerplate code - wrapping each function call will be ugly
+    - no option to just annotate main function (procedural macros could achieve this, but they do some compile-time modifications)
+    - potentially more complex
+
+An example pseudo-implementation of this could look like this:
+```rust
+fn skip_if_done<F>(func: F)
+where
+    F: Fn() + 'static,
+{
+    // Decorator logic here
+    // Check the state file and decide whether to execute the function or not
+
+    if should_execute {
+        func();
+        // write func outputs to state file
+    } else {
+        // Do nothing or return an appropriate value
+    }
+}
+
+fn main() {
+    skip_if_done(|| {
+        dispatcher.send_eth(...);
+    });
+}
+```
+
+- Use a procedural macro to automatically apply decorator-like behaviour to functions, which will decide on whether
+given function should be executetd or not (based on state file)
+  - pros:
+    - elegant solution
+    - could be applied selectively (per function) or globally (to main)
+  - cons:
+    - more complex solution
+    - probably not very suitable for runtime based behaviours (like reading json file) - requires more research
+
+An example pseudo-implementation of this could look like this:
+```rust
+// skip_if_done.rs
+use proc_macro_hack::proc_macro_hack;
+
+#[proc_macro_hack]
+pub use skip_if_done_impl::skip_if_done;
+
+
+// skip_if_done_impl.rs
+use quote::quote;
+use std::env;
+
+fn should_execute() -> bool {
+    // Your logic to check whether the function should be executed goes here
+    true
+}
+
+#[macro_export]
+macro_rules! skip_if_done {
+    ($($tokens:tt)*) => {
+        {
+            if should_execute() {
+                $($tokens)*
+            }
+        }
+    };
+}
+
+// main.rs
+#[path = "skip_if_done.rs"]
+mod skip_if_done;
+
+fn main() {
+    skip_if_done! {
+        dispatcher.send_eth(...);
+    }
+}
+```
+
+Given all this, the most obvious solutions seem to be:
+  - for cast subcommands functions:
+    - adjust them to handle checking the file
+  - for dispatcher functions:
+    - implement a higher order function
+
+Picking those would ensure us that:
+  - syntax is relatively boiler-code free (only dispatcher functions are annotated)
+  - idempotency is achieved
+  - it is quite easy to write a script
+
+### example script
+An example deployment script could look like this:
+
+```cairo
+// we might want to rename account functions to avoid confusion
+use cast::starknet_commands::account::create::create as create_account
+(...)
+
+fn make_account(...) -> Result<InvokeResponse> {
+  let user = create_account(...);
+  let prefund_account = invoke(...);
+  deploy_account(...);
+}
+
+fn main() {
+  let user = make_account(...);
+  skip_if_done(|| {
+        dispatcher.increase_balance(...);
+    });
+  let get_balance = skip_if_done(|| {
+        dispatcher.get_balance(...);
+    });
```

Comment:
What would happen if get_balance was called? Will it return an old value? What if the state changed between calls?

Reply:
If this is called with 'latest' block, that would potentially change every next function that depends on this value, so they will be done again (as they are treated as "new").
If it was called with fixed block number, this will always return the same value (correct me if I'm wrong), so nothing changes.

Reply:
Ok, but if it was done once and is marked with `skip_if_done` would it be called again?

Reply:
if marked with skip_if_done, it won't be called again. to have it called again, you'd need to "un-mark" it. Do you think this behaviour should change?

Reply:
I think marking `calls` (not `invokes`) should be disabled entirely. Maybe not if the block is specified. 
I can think of a situation where:
- someone calls a contract and wants to use this value in the next call
- something is wrong with the script and it fails
- in the meantime state changes
- someone runs the script but the value is already `"fetched"`
- wrong value is saved on the chain

Reply:
To be clear - I think skipping should be allowed for calls with fixed blocks (not latest/pending). I see that pending/latest blocks can be problematic. Your concerns are very valid.

However, wouldn't it be too limiting for some people? It all may be remedied with just not using `skip_if_done` for certain calls. And keeping it the way it is currently proposed, it would allow users to have a script that can actually update on chain state based on some external dependency (call). I can imagine such behaviour would be desired by _some_ people. I'd like to provide the users with options, and let them decide.

I am aware that that could prove to be problematic, or even confusing to some. If we go that way, it is crucial to make it really clear what is the behaviour, and what users should do to achieve a behaviour they desire. This owuld have to be described very well in the docs.

Any thoughts?

Reply:
I agree that options are overall better but as you said if we go that way the docs would have to be `very` clear.
We can stick to your proposal but I think sometimes skipping the `calls` can lead to unexpected behaviour. We should definitely include that in the docs.